arduino/smartcar/smartcar.ino

```diff
@@ -14,18 +22,39 @@ void setup() {
 }
 
 void loop() {
-  car.setSpeed(90);
+  car.setSpeed(70); 
   preventCrash();
 }
 
-void preventCrash() {
-  unsigned int fDistance = front.getDistance();
-  if (fDistance > 0 && fDistance < 200) {
+void preventCrash(){
+  preventCrashWithUltrasonic(); 
+  preventCrashWithInfrared(frontIR, 0, 40); // Helps in short distances object detection.
+}
+
+void preventCrashWithUltrasonic() {
+  unsigned int distanceUS = frontUS.getDistance(); // min distance: 0.2 m, MAX distance: 4 m.
+  if (distanceUS > 20 && distanceUS < 50) {
+    Serial.print(frontUS.getDistance());
+    Serial.println(" Obstacle detected by ultrasonic sensor. ");
     car.setSpeed(0);
     delay(500);
-    control.overrideMotorSpeed(100, -100);
+    car.overrideMotorSpeed(100, -100);
     delay(800);
     car.setSpeed(0);
     delay(200);
   }
 }
+
+void preventCrashWithInfrared(DistanceSensor& distanceSensor, float minDistance, float maxDistance) { 
```

Comment:
Since you are taking any instance of a class implementing the interface DistanceSensor you can reuse this method for the ultrasonic, since SR04 implements DistanceSensor as well. That way you don't have two methods using very similar code. I also don't see a point in taking minDistance as an argument rather than hardcoding it as 0.

Reply:
I have tried that. Does not work.

Reply:
I fetched your branch and tried it myself. Both my points work just fine.